File: tools/testing/connection/httpx/postgres.py
#!/usr/bin/env python3
"""
PostgreSQL Connection Test (using asyncpg/psycopg)

Authentication: Password
Protocol: PostgreSQL wire protocol (not HTTP)
Health Check: SELECT 1

Note: PostgreSQL uses its own wire protocol, not HTTP.
This file demonstrates both asyncpg and psycopg connections.
"""

import asyncio
import os
from typing import Any

# ============================================================================
# Configuration - Override these values
# ============================================================================

CONFIG = {
    # Required
    "POSTGRES_HOST": os.getenv("POSTGRES_HOST", "localhost"),
    "POSTGRES_PORT": int(os.getenv("POSTGRES_PORT", "5432")),
    "POSTGRES_USER": os.getenv("POSTGRES_USER", "postgres"),
    "POSTGRES_PASSWORD": os.getenv("POSTGRES_PASSWORD", "your_password_here"),
    "POSTGRES_DB": os.getenv("POSTGRES_DB", "postgres"),

    # Optional
    "POSTGRES_SCHEMA": os.getenv("POSTGRES_SCHEMA", "public"),

    # Connection URL (alternative)
    "DATABASE_URL": os.getenv("DATABASE_URL", ""),
}


def get_connection_url() -> str:
    """Get connection URL."""
    if CONFIG["DATABASE_URL"]:
        return CONFIG["DATABASE_URL"]
    return (
        f"postgresql://{CONFIG['POSTGRES_USER']}:{CONFIG['POSTGRES_PASSWORD']}"
        f"@{CONFIG['POSTGRES_HOST']}:{CONFIG['POSTGRES_PORT']}/{CONFIG['POSTGRES_DB']}"
    )


# ============================================================================
# Health Check (using asyncpg)
# ============================================================================

async def health_check_asyncpg() -> dict[str, Any]:
    """Perform health check using asyncpg."""
    print("\n=== PostgreSQL Health Check (asyncpg) ===\n")

    try:
        import asyncpg

        print(f"Connecting to: {CONFIG['POSTGRES_HOST']}:{CONFIG['POSTGRES_PORT']}/{CONFIG['POSTGRES_DB']}")

        conn = await asyncpg.connect(
            host=CONFIG["POSTGRES_HOST"],
            port=CONFIG["POSTGRES_PORT"],
            user=CONFIG["POSTGRES_USER"],
            password=CONFIG["POSTGRES_PASSWORD"],
            database=CONFIG["POSTGRES_DB"],
        )

        # Test connection
        result = await conn.fetchval("SELECT 1")
        print(f"SELECT 1: {result}")

        # Get version
        version = await conn.fetchval("SELECT version()")
        print(f"Version: {version}")

        # Get current database
        db = await conn.fetchval("SELECT current_database()")
        print(f"Database: {db}")

        await conn.close()

        return {"success": True, "data": {"version": version, "database": db}}
    except ImportError:
        print("Error: asyncpg package not installed. Run: pip install asyncpg")
        return {"success": False, "error": "asyncpg package not installed"}
    except Exception as e:
        print(f"Error: {e}")
        return {"success": False, "error": str(e)}


# ============================================================================
# Health Check (using psycopg3)
# ============================================================================

async def health_check_psycopg() -> dict[str, Any]:
    """Perform health check using psycopg3."""
    print("\n=== PostgreSQL Health Check (psycopg3) ===\n")

    try:
        import psycopg

        print(f"Connecting to: {CONFIG['POSTGRES_HOST']}:{CONFIG['POSTGRES_PORT']}/{CONFIG['POSTGRES_DB']}")

        async with await psycopg.AsyncConnection.connect(
            host=CONFIG["POSTGRES_HOST"],
            port=CONFIG["POSTGRES_PORT"],
            user=CONFIG["POSTGRES_USER"],
            password=CONFIG["POSTGRES_PASSWORD"],
            dbname=CONFIG["POSTGRES_DB"],
        ) as conn:
            async with conn.cursor() as cur:
                # Test connection
                await cur.execute("SELECT 1")
                result = await cur.fetchone()
                print(f"SELECT 1: {result[0]}")

                # Get version
                await cur.execute("SELECT version()")
                version = (await cur.fetchone())[0]
                print(f"Version: {version}")

                # Get current database
                await cur.execute("SELECT current_database()")
                db = (await cur.fetchone())[0]
                print(f"Database: {db}")

        return {"success": True, "data": {"version": version, "database": db}}
    except ImportError:
        print("Error: psycopg package not installed. Run: pip install psycopg[binary]")
        return {"success": False, "error": "psycopg package not installed"}
    except Exception as e:
        print(f"Error: {e}")
        return {"success": False, "error": str(e)}


# ============================================================================
# Sample Operations
# ============================================================================

async def sample_operations() -> dict[str, Any]:
    """Perform sample PostgreSQL operations."""
    print("\n=== Sample PostgreSQL Operations ===\n")

    try:
        import asyncpg

        conn = await asyncpg.connect(
            host=CONFIG["POSTGRES_HOST"],
            port=CONFIG["POSTGRES_PORT"],
            user=CONFIG["POSTGRES_USER"],
            password=CONFIG["POSTGRES_PASSWORD"],
            database=CONFIG["POSTGRES_DB"],
        )

        # List schemas
        schemas = await conn.fetch("""
            SELECT schema_name
            FROM information_schema.schemata
            ORDER BY schema_name
        """)
        print("Schemas:")
        for schema in schemas[:10]:
            print(f"  - {schema['schema_name']}")

        # List tables in public schema
        tables = await conn.fetch("""
            SELECT table_name
            FROM information_schema.tables
            WHERE table_schema = $1
            ORDER BY table_name
        """, CONFIG["POSTGRES_SCHEMA"])
        print(f"\nTables in {CONFIG['POSTGRES_SCHEMA']}:")
        for table in tables[:10]:
            print(f"  - {table['table_name']}")

        # Get database size
        size = await conn.fetchval("""
            SELECT pg_size_pretty(pg_database_size(current_database()))
        """)
        print(f"\nDatabase size: {size}")

        await conn.close()

        return {"success": True}
    except ImportError:
        print("Error: asyncpg package not installed. Run: pip install asyncpg")
        return {"success": False, "error": "asyncpg package not installed"}
    except Exception as e:
        print(f"Error: {e}")
        return {"success": False, "error": str(e)}


# ============================================================================
# HTTP-based PostgreSQL (PostgREST example)
# ============================================================================

async def health_check_postgrest() -> dict[str, Any]:
    """Health check for PostgREST API."""
    print("\n=== PostgREST API Health Check ===\n")

    import httpx

    postgrest_url = os.getenv("POSTGREST_URL", "")
    postgrest_token = os.getenv("POSTGREST_TOKEN", "")

    if not postgrest_url:
        print("PostgREST URL not configured. Set POSTGREST_URL environment variable")
        return {"success": False, "error": "PostgREST URL not configured"}

    async with httpx.AsyncClient() as client:
        try:
            headers = {"Accept": "application/json"}
            if postgrest_token:
                headers["Authorization"] = f"Bearer {postgrest_token}"

            response = await client.get(
                f"{postgrest_url}/",
                headers=headers,
            )

            print(f"Status: {response.status_code}")
            if response.status_code == 200:
                print("PostgREST is accessible")
            else:
                print(f"Response: {response.text}")

            return {"success": response.status_code == 200}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


# ============================================================================
# Run Tests
# ============================================================================

async def main():
    """Run connection tests."""
    print("PostgreSQL Connection Test")
    print("==========================")
    print(f"Host: {CONFIG['POSTGRES_HOST']}:{CONFIG['POSTGRES_PORT']}")
    print(f"Database: {CONFIG['POSTGRES_DB']}")
    print(f"User: {CONFIG['POSTGRES_USER']}")

    await health_check_asyncpg()
    # await health_check_psycopg()  # Alternative using psycopg3
    # await sample_operations()
    # await health_check_postgrest()  # For PostgREST HTTP API


if __name__ == "__main__":
    asyncio.run(main())

File: tools/testing/connection/undici/postgres.mjs
/**
 * PostgreSQL - Connection Test (using pg)
 *
 * Note: PostgreSQL uses its own protocol, not HTTP.
 * This file demonstrates connection testing with pg (node-postgres).
 *
 * Authentication: Password
 * Default Port: 5432
 * Health Query: SELECT 1
 *
 * TLS/SSL Options:
 *   NODE_TLS_REJECT_UNAUTHORIZED=0  - Ignore all certificate errors
 *   REQUEST_CA_BUNDLE=/path/to/ca   - Custom CA bundle file
 *   SSL_CERT_FILE=/path/to/cert     - Custom SSL certificate file
 *   NODE_EXTRA_CA_CERTS=/path/to/ca - Additional CA certificates
 */

import pg from 'pg';
import { readFileSync } from 'node:fs';
const { Client, Pool } = pg;

// ============================================================================
// Configuration - Override these values
// ============================================================================

const CONFIG = {
  // Option 1: Full Connection URL
  DATABASE_URL: process.env.DATABASE_URL || '',

  // Option 2: Individual Components
  POSTGRES_HOST: process.env.POSTGRES_HOST || 'localhost',
  POSTGRES_PORT: parseInt(process.env.POSTGRES_PORT || '5432', 10),
  POSTGRES_USER: process.env.POSTGRES_USER || 'postgres',
  POSTGRES_PASSWORD: process.env.POSTGRES_PASSWORD || '',
  POSTGRES_DB: process.env.POSTGRES_DB || 'postgres',

  // Optional: SSL Configuration
  POSTGRES_SSLMODE: process.env.POSTGRES_SSLMODE || 'prefer',
  // Options: disable, allow, prefer, require, verify-ca, verify-full

  // Optional: TLS Configuration
  // Set to false to ignore certificate errors (default: false for testing)
  // NODE_TLS_REJECT_UNAUTHORIZED=0, REQUEST_CA_BUNDLE=null, SSL_CERT_FILE=null, NODE_EXTRA_CA_CERTS=null
  REJECT_UNAUTHORIZED: false,

  // Optional: Custom CA certificates (disabled by default for testing)
  CA_BUNDLE: null,
};

// ============================================================================
// TLS Configuration Helper
// ============================================================================

function getTlsOptions() {
  const options = {
    rejectUnauthorized: CONFIG.REJECT_UNAUTHORIZED,
  };

  // Load custom CA bundle if specified
  if (CONFIG.CA_BUNDLE) {
    try {
      options.ca = readFileSync(CONFIG.CA_BUNDLE).toString();
      console.log(`Using custom CA bundle: ${CONFIG.CA_BUNDLE}`);
    } catch (err) {
      console.warn(`Warning: Could not read CA bundle: ${err.message}`);
    }
  }

  return options;
}

// ============================================================================
// Create PostgreSQL Client
// ============================================================================

function createClient() {
  const tlsOptions = getTlsOptions();

  if (CONFIG.DATABASE_URL) {
    console.log(`Using connection URL: ${CONFIG.DATABASE_URL.replace(/:[^:@]+@/, ':***@')}`);
    return new Client({
      connectionString: CONFIG.DATABASE_URL,
      ssl: CONFIG.DATABASE_URL.includes('sslmode=') ? undefined : tlsOptions,
    });
  }

  const sslConfig = getSslConfig(tlsOptions);

  return new Client({
    host: CONFIG.POSTGRES_HOST,
    port: CONFIG.POSTGRES_PORT,
    user: CONFIG.POSTGRES_USER,
    password: CONFIG.POSTGRES_PASSWORD,
    database: CONFIG.POSTGRES_DB,
    ssl: sslConfig,
  });
}

function getSslConfig(tlsOptions) {
  switch (CONFIG.POSTGRES_SSLMODE) {
    case 'disable':
      return false;
    case 'require':
      return {
        ...tlsOptions,
        rejectUnauthorized: false,
      };
    case 'verify-ca':
    case 'verify-full':
      return tlsOptions;
    case 'allow':
    case 'prefer':
    default:
      return {
        ...tlsOptions,
        rejectUnauthorized: false,
      };
  }
}

// ============================================================================
// Health Check
// ============================================================================

async function healthCheck() {
  console.log('\n=== PostgreSQL Health Check ===\n');

  const client = createClient();

  try {
    await client.connect();
    console.log('Connected successfully!');

    const result = await client.query('SELECT 1 as health_check');

    console.log(`Query Result: ${JSON.stringify(result.rows[0])}`);

    return { success: true, result: result.rows[0] };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await client.end();
  }
}

// ============================================================================
// Sample Queries
// ============================================================================

async function getVersion() {
  console.log('\n=== PostgreSQL Version ===\n');

  const client = createClient();

  try {
    await client.connect();

    const result = await client.query('SELECT version()');

    console.log(`Version: ${result.rows[0].version}`);

    return { success: true, version: result.rows[0].version };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await client.end();
  }
}

async function listDatabases() {
  console.log('\n=== List Databases ===\n');

  const client = createClient();

  try {
    await client.connect();

    const result = await client.query(`
      SELECT datname FROM pg_database
      WHERE datistemplate = false
      ORDER BY datname
    `);

    console.log(`Found ${result.rows.length} databases`);
    result.rows.forEach((row) => {
      console.log(`  - ${row.datname}`);
    });

    return { success: true, databases: result.rows };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await client.end();
  }
}

async function listTables(schema = 'public') {
  console.log(`\n=== List Tables (${schema}) ===\n`);

  const client = createClient();

  try {
    await client.connect();

    const result = await client.query(`
      SELECT tablename FROM pg_tables
      WHERE schemaname = $1
      ORDER BY tablename
    `, [schema]);

    console.log(`Found ${result.rows.length} tables`);
    result.rows.forEach((row) => {
      console.log(`  - ${row.tablename}`);
    });

    return { success: true, tables: result.rows };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await client.end();
  }
}

async function getDatabaseSize() {
  console.log('\n=== Database Size ===\n');

  const client = createClient();

  try {
    await client.connect();

    const result = await client.query(`
      SELECT pg_size_pretty(pg_database_size(current_database())) as size
    `);

    console.log(`Database size: ${result.rows[0].size}`);

    return { success: true, size: result.rows[0].size };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await client.end();
  }
}

async function getConnections() {
  console.log('\n=== Active Connections ===\n');

  const client = createClient();

  try {
    await client.connect();

    const result = await client.query(`
      SELECT count(*) as connections FROM pg_stat_activity
    `);

    console.log(`Active connections: ${result.rows[0].connections}`);

    return { success: true, connections: result.rows[0].connections };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await client.end();
  }
}

// ============================================================================
// Run Tests
// ============================================================================

async function main() {
  console.log('PostgreSQL Connection Test');
  console.log('==========================');
  console.log(`Host: ${CONFIG.POSTGRES_HOST}:${CONFIG.POSTGRES_PORT}`);
  console.log(`Database: ${CONFIG.POSTGRES_DB}`);
  console.log(`User: ${CONFIG.POSTGRES_USER}`);
  console.log(`SSL Mode: ${CONFIG.POSTGRES_SSLMODE}`);
  console.log(`TLS Verify: ${CONFIG.REJECT_UNAUTHORIZED}`);
  console.log(`CA Bundle: ${CONFIG.CA_BUNDLE || 'System default'}`);

  await healthCheck();
  // await getVersion();
  // await listDatabases();
  // await listTables();
  // await getDatabaseSize();
  // await getConnections();
}

main().catch(console.error);
